model_templates/gpu_nemo_textgen/custom_model/custom.py

```diff
@@ -1,25 +1,101 @@
+import io
 import os
 import boto3
 from mlpiper.extra.aws_helper import AwsHelper
-from datarobot_drum import RuntimeParameters
+from pathlib import Path
 
+from datarobot_drum import RuntimeParameters
+from datarobot_drum.drum.enum import LOGGER_NAME_PREFIX
 
 # this path is required by NeMo Inference Server
 NEMO_MODEL_STORE_DIR = "/model-store"
 
 
 def load_model(code_dir: str):
-    print("Downloading model from S3...")
+    load_model_response = "succeeded"  # a non-empty response is required to signal that load_model succeeded
+
+    s3_url = RuntimeParameters.get("s3Url", "")
+    if s3_url:
+        s3_credential = RuntimeParameters.get("s3Credential")
+        s3_client = S3Client(s3_url, s3_credential)
+
+        keys, dirs = s3_client.list_objects()
+        s3_client.download_files(NEMO_MODEL_STORE_DIR, keys, dirs)
+        return load_model_response
+
+    ngc_registry_url = RuntimeParameters.get("ngcRegistryUrl", "")
+    if ngc_registry_url:
+        ngc_credential = RuntimeParameters.get("ngcCredential")
+        ngc_client = NGCRegistryClient(ngc_credential)
+        ngc_client.download_model_version(ngc_registry_url)
+        return load_model_response
+
+    raise Exception(
+        "Can't download model artifacts. The 'ngcRegistryUrl' or 's3Url'"
+        " are expected to be set in the Runtime Parameters."
+    )
+
+
+class NGCRegistryClient:
+    logger = logging.getLogger(LOGGER_NAME_PREFIX + "." + __name__)
+
+    def __init__(self, ngc_credential):
+        if ngc_credential["credentialType"] != SecretType.API_TOKEN.value:
+            raise ValueError("NGC credential is expected to be of type 'API_TOKEN'.")
+        self.api_token = ngc_credential["apiToken"]
+        self._configure_ngc_client()
+
+    def download_model_version(self, ngc_registry_url):
+        ngc_org, ngc_team = self._parse_ngc_registry_url(ngc_registry_url)
+
+        cmd = [
+            "ngc",
+            "registry",
+            "model",
+            "download-version",
+            "--org",
+            ngc_org,
+            "--team",
+            ngc_team,
+            ngc_registry_url,
+        ]
+
+        p = subprocess.Popen(
+            cmd,
+            env=os.environ,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+        )
+        stdout, stderr = p.communicate()
+
+        if len(stdout):
+            self.logger.info(stdout)
+        if len(stderr):
+            self.logger.error(stderr)
+
+        assert p.returncode == 0, "Failed to download model version from NGC registry"
+
```

Comment:
Wouldn't this be simplier to just call `subprocess.run(cmd, ..., check=True)`? You can capture output or not if you wish too.

Reply:
You may need to catch the subprocess thrown exception and raise a the proper drum exception because I don't think generic exceptions are properly output in the custom-models infra? But if you have `capture_output=True` all the stdout and stderr should be included in the exception.

Reply:
- changed code to use the `subprocess.run`
- raising the `DrumCommonException`
https://github.com/datarobot/datarobot-user-models/blob/b978bde13f8362d197835f0c42acaaa05e8f055d/custom_model_runner/datarobot_drum/drum/gpu_predictors/utils.py#L74-L85